cpp/include/raft/sparse/hierarchy/detail/agglomerative.cuh

```diff
@@ -0,0 +1,366 @@
+
+  // 2. Run union-find in parallel for each pair of folds
+
+  // 3. Sort individual label hierarchies
+
+  // 4. Merge label hierarchies together
+}
+
+template <typename value_idx>
+__global__ void write_levels_kernel(const value_idx *children,
+                                    value_idx *parents, value_idx n_vertices) {
+  value_idx tid = blockDim.x * blockIdx.x + threadIdx.x;
+  if (tid < n_vertices) {
+    value_idx level = tid / 2;
+    value_idx child = children[tid];
+    parents[child] = level;
+  }
+}
+
+/**
+ * Instead of propagating a label from roots to children,
+ * the children each iterate up the tree until they find
+ * the label of their parent. This increases the potential
+ * parallelism.
+ * @tparam value_idx
+ * @param children
+ * @param parents
+ * @param n_leaves
+ * @param labels
+ */
+template <typename value_idx>
+__global__ void inherit_labels(const value_idx *children,
+                               const value_idx *levels, size_t n_leaves,
+                               value_idx *labels, int cut_level,
+                               value_idx n_vertices) {
+  value_idx tid = blockDim.x * blockIdx.x + threadIdx.x;
+
+  if (tid < n_vertices) {
+    value_idx node = children[tid];
+    value_idx cur_level = tid / 2;
+
+    /**
+     * Any roots above the cut level should be ignored.
+     * Any leaves at the cut level should already be labeled
+     */
+    if (cur_level > cut_level) return;
+
+    value_idx cur_parent = node;
+    value_idx label = labels[cur_parent];
+
+    while (label == -1) {
+      cur_parent = cur_level + n_leaves;
+      cur_level = levels[cur_parent];
+      label = labels[cur_parent];
+    }
+
+    labels[node] = label;
+  }
+}
+
+template <typename value_idx>
+struct init_label_roots {
+  init_label_roots(value_idx *labels_) : labels(labels_) {}
+
+  template <typename Tuple>
+  __host__ __device__ void operator()(Tuple t) {
+    labels[thrust::get<1>(t)] = thrust::get<0>(t);
+  }
+
+ private:
+  value_idx *labels;
+};
+
+/**
+ * Cuts the dendrogram at a particular level where the number of nodes
+ * is equal to n_clusters, then propagates the resulting labels
+ * to all the children.
+ *
+ * @tparam value_idx
+ * @param handle
+ * @param labels
+ * @param children
+ * @param n_clusters
+ * @param n_leaves
+ */
+template <typename value_idx, int tpb = 256>
+void extract_flattened_clusters(const raft::handle_t &handle, value_idx *labels,
+                                const value_idx *children, size_t n_clusters,
+                                size_t n_leaves) {
+  auto d_alloc = handle.get_device_allocator();
+  auto stream = handle.get_stream();
+
+  /**
+   * Compute levels for each node
+   *
+   *     1. Initialize "levels" array of size n_leaves * 2
+   *
+   *     2. For each entry in children, write parent
+   *        out for each of the children
+   */
+
+  size_t n_edges = (n_leaves - 1) * 2;
+
+  thrust::device_ptr<value_idx> d_ptr =
```

Comment:
Thrust functions will run the host path when a pointer to device memory is used directly. see: https://docs.nvidia.com/cuda/thrust/index.html#iterators-and-static-dispatching.

Unfortunately, while we want the function signature here to let the caller know we don't intend on modifying `chilren`, I get a compile error when trying to use `device_ptr_cast` on const, so I'm casting the const away temporarily for purposes of the computation. If you know a better way to do this then I'd be willing to try it out.

Reply:
That's why I mentioned about using `rmm::exec_policy` which is a device policy and uses RMM's allocators. So I think device ptr's are unneeded. As for const, you can also do `thrust::device_ptr<const value_idx>` (for future purposes)

Reply:
This causes a segfault:
```
  auto thrust_policy = rmm::exec_policy(stream);
  value_idx n_vertices = *(thrust::max_element(thrust_policy,
                                               children, children + n_edges)) +
                         1;
```

This does not:

```
  auto thrust_policy = rmm::exec_policy(stream);
  thrust::device_ptr<const value_idx> d_ptr =
    thrust::device_pointer_cast(children);
  value_idx n_vertices = *(thrust::max_element(thrust_policy,
                                               d_ptr, d_ptr + n_edges)) +
                         1;
```

Reply:
That's weird, I have used raw pointers with thrust without any issues. Example here: https://github.com/rapidsai/raft/blob/branch-0.19/cpp/include/raft/sparse/mst/detail/mst_solver_inl.cuh#L86

Reply:
Can you try using this with an example that does not use `thrust::max_element`? Just to confirm with our suspicions of a bug in min-max and maybe that causing the deref there to cause a seg-fault

Reply:
This one is not compiling: 
```
template <typename value_idx, typename value_t>
void sort_coo_by_data(value_idx *rows, value_idx *cols, value_t *data,
                      value_idx nnz, cudaStream_t stream) {
  auto thrust_policy = rmm::exec_policy(stream);
  thrust::device_ptr<value_idx> t_rows = thrust::device_pointer_cast(rows);
  thrust::device_ptr<value_idx> t_cols = thrust::device_pointer_cast(cols);
  thrust::device_ptr<value_t> t_data = thrust::device_pointer_cast(data);

  auto first = thrust::make_zip_iterator(thrust::make_tuple(t_rows, t_cols));

  thrust::sort_by_key(thrust_policy, data, data + nnz, first);
}
```

I see the benefit of using the exec_policy but iI'm really failing to see the benefit here when the code works with the pointer casts. Is it just because you're not in love with the extra lines of code? I can't immediately think of a potential pitfall for having these in there.

Reply:
Yeah, not a big fan of the extra lines of code for what I have seen to work in other places without an issue. Let's sync up offline on this and I can take a stab at compiling an example by myself. Which in this case, I'd try to have the tuple not be composed of device ptr's either, but no point in me trying to try out stuff to see if it works, when you can move ahead with more important stuff. I can just try it out myself and see if it works and update you on that

Reply:
This doesn't compile either, unfortunately.

```
template <typename value_idx, typename value_t>
void sort_coo_by_data(value_idx *rows, value_idx *cols, value_t *data,
                      value_idx nnz, cudaStream_t stream) {
  auto thrust_policy = rmm::exec_policy(stream);
  auto first = thrust::make_zip_iterator(thrust::make_tuple(rows, cols));

  thrust::sort_by_key(thrust_policy, data, data + nnz, first);
}
```

There are places which have worked without explicit pointer casts. I'm okay making this change in those places and adding the rmm execution policy everywhere but I'm not completely on board w/ setting that as a general rule just for a few extra lines of code that don't seem to have any other adverse affects. 

I'll push what I have if you'd like to play around with it.

Reply:
Yep, that's a fair middle ground here for something that *should* work out of the box, but if it's not then we don't waste too much time on it for sure.